lib/index.js

```diff
@@ -14,110 +15,74 @@ const DEFAULT_OPTIONS = {
     string: '\x1b[32m',
     special: '\x1b[33m',
     bracket: '\x1b[33m',
+    comment: '\x1b[2m\x1b[90m',
     clear: '\x1b[0m'
   }
 }
 
-const SPLIT_CHARS = '[^a-zA-Z_]'
-
 const DEFAULT_KEYWORD = 'default'
 
 const highlighters = [
-  {
-    name: 'keyword',
-    group: 1,
-    regex: new RegExp(`(^|${SPLIT_CHARS})(${keywords.join('|')})(?=${SPLIT_CHARS}|$)`, 'gi')
-  },
-  {
-    name: 'special',
-    regex: /(=|!=|%|\/|\*|-|,|;|:|\+|<|>)/g
-  },
-  {
-    name: 'function',
-    regex: /(\w+?)\(/g,
-    trimEnd: 1
-  },
-  {
-    name: 'number',
-    regex: /(\b\d+(?:\.\d+)?)/g
-  },
-  {
-    name: 'string',
-    regex: /(['](?:\\'|.)*?[']|["](?:\\"|.)*?["]|[`](?:\\`|.)*?[`])/g
-  },
-  {
-    name: 'bracket',
-    regex: /([()])/g
-  }
-]
+  /\b(?<number>  \d+ (?:\.\d+)?  )\b/,
 
-function getSegments (sqlString) {
-  const matches = []
-
-  for (const hl of highlighters) {
-    let match
-
-    // This is probably the one time when an assignment inside a condition makes sense
-    // eslint-disable-next-line no-cond-assign
-    while (match = hl.regex.exec(sqlString)) {
-      let text = match[0]
-      let boringLength = 0
-
-      // If a specific group is requested, use that group instead, and make sure
-      // we offset the index by the length of the preceding groups
-      if (hl.group) {
-        text = match[hl.group + 1]
-        for (let i = 1; i <= hl.group; i++) {
-          boringLength += match[i].length
-        }
-      }
+  // Note: Repeating string escapes like 'sql''server' will also work as they are just repeating strings
+  /(?<string>  '(?: [^'\\] | \\. )*'  |  "(?: [^"\\] | \\. )*"  |  `(?: [^`\\] | \\. )*`  )/,
 
-      const trimmedText = hl.trimEnd
-        ? text.substring(0, text.length - hl.trimEnd)
-        : text
-      matches.push({
-        name: hl.name,
-        start: match.index + boringLength,
-        length: trimmedText.length
-      })
-    }
-  }
+  /(?<comment>  --[^\n\r]*  |  #[^\n\r]*  |  \/\*  (?: [^*] | \* (?!\/) )* \*\/  )/,
+
+  // Future improvement: Comments should be allowed between the function name and the opening parenthesis
+  /\b(?<function> \w+ ) (?= \s*\( )/,
+
+  /(?<bracket> [()] )/,
+
+  /(?<special> != | [=%*/\-+,;:<>] )/
+]
 
-  const sortedMatches = matches.slice().sort((a, b) => a.start - b.start)
+function getRegexString (regex) {
+  const str = regex.toString()
+  return str.replace(/^\/|\/\w*$|[\t ]+/g, '')
+}
+
+// Regex of the shape /(.*?)|((?<token1>...)|(?<token2>...)|...|$)/y
+const tokenizer = new RegExp(
+  '(.*?)(' +
+  '\\b(?<keyword>' + keywords.join('|') + ')\\b|' +
```

Comment:
Would it be possible to move the keywords highlighter into the `highlighters` array? I think it makes more sense to keep them together and not pollute this regex.

Reply:
That would do unnecessary work of `.toString()`ing that regex and parsing it again. And since the keywords regex is so big big I would want to avoid that.

Also this is where the actually used regex is constructed, with several additional parts, so imo it doesn't seem too far off.

But I could just move the keywords regex into its own variable. Like `const keywordsHighlighter = '...'` and put it just above `const highlighters = ...`, which could more clear. What do you think?

Reply:
> unnecessary work of .toString()ing that regex and parsing it again

If we go to that way, better remove the `join('|')`, and keywords must be a pipe separated string

Reply:
Let's keep this format